flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java

```diff
@@ -196,16 +217,21 @@ public SlotReport createSlotReport(ResourceID resourceId) {
 	public boolean allocateSlot(int index, JobID jobId, AllocationID allocationId, Time slotTimeout) {
 		checkInit();
 
+		Preconditions.checkArgument(index < numberSlots);
+
 		TaskSlot taskSlot = allocationIDTaskSlotMap.get(allocationId);
 		if (taskSlot != null) {
 			LOG.info("Allocation ID {} is already allocated in {}.", allocationId, taskSlot);
 			return false;
 		}
-		taskSlot = taskSlots.get(index);
+
+		taskSlot = new TaskSlot(index, defaultSlotResourceProfile, memoryPageSize);
 
 		boolean result = taskSlot.allocate(jobId, allocationId);
 
 		if (result) {
+			taskSlots.put(index, taskSlot);
```

Comment:
True, `result` should always be `true`. The idea here is to make `TaskSlotTable` independent from the implementation of `TaskSlot`, only stick to the its contract (which is that `TaskSlot#allocate` might fail).

Agree that we should do sanity check on the duplicated `index`.

Reply:
As discussed offline, I'll remove `TaskSlotState#FREE` (task slot will be created with the state `ALLOCATED`), and `TaskSlot#allocate` as well (because task slot can no longer be allocated once created).